### ISSUE TITLE
Can users move generated computation  graph manually in the forward pass from GPU to CPU? And then back to GPU for backward

### ISSUE THREAD

**Issue by xcwanAndy** [pytorch/pytorch#57100]
## 🚀 Feature
I wonder if PyTorch allows such feature: users can manually move the forward generated computation graph from GPU to CPU, and then back to GPU for backward?

## Motivation
I have such feature request mainly due to: 

I have a demand to slice a large amount of data into batches and then finish all of them in the forward pass, and afterwards backwarding them (Yes, I __cannot__ use mini-batch training in my task).  However, as such large data cannot fit into GPU memory, I have to move the generated computation graph for a specific batch into CPU memory. And after all forward tasks finished, send them one-by-one back to GPU for backward tasks.

## Pitch
After this feature, users can move generated computation  graph manually in the forward pass from GPU to CPU, and then back to GPU for backward.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano

**Comment by zou3519** [comment]
This seems like a use case that more folks would have

**Comment by xcwanAndy** [comment]
@zou3519 Yes, thanks a lot for your comment! I wonder if right now PyTorch has provided such APIs ? If not, is there any quick ways for me to hack into __autograd__ ? As I'm in a hurry for such feature.

Thanks a lot for your help!

**Comment by albanD** [comment]
Hi,

There is no such feature right now no :/ 
The graph is pretty deep in c++ I'm afraid.

**Comment by xcwanAndy** [comment]
@albanD Would there be any plan for this feature? I guess it's becoming popular.

**Comment by albanD** [comment]
Hi,

This is definitely something we would like to have in one form or another. See also https://github.com/pytorch/pytorch/issues/28997 for a similar request.
The main problem is that there is no obvious API for this and no simple implementation. So that would be a fairly big project to add.

**Comment by xcwanAndy** [comment]
@Varal7 Thanks for helping with this issue. May I ask if the API is available now to use? I'd like to compile the source code to manually operate the computation graph.

**Comment by Varal7** [comment]
Hi @xcwanAndy,
The feature is not ready to use. If you want to try it, you can build Pytorch from this branch: https://github.com/pytorch/pytorch/tree/gh/varal7/18/head.

**Comment by xcwanAndy** [comment]
@Varal7 Thanks a lot! I'll try this branch now.

**Comment by xcwanAndy** [comment]
@albanD Hi, I wonder if `torch.autograd.graph.save_on_cpu()` can also save the generated computation graph from GPU to CPU, such that I can manually operate on both the input/output tensors and the computation graph.

**Comment by Varal7** [comment]
Hi @xcwanAndy, you can use `torch.autograd.graph.save_on_cpu()` to offload the tensors saved by the computation onto CPU, as described by https://pytorch.org/docs/stable/autograd.html?#torch.autograd.graph.save_on_cpu.

To manually traverse the graph, you can check out the autograd docs about [how this graph is computed](https://pytorch.org/docs/master/notes/autograd.html#how-autograd-encodes-the-history) and refer to how [torchviz](https://github.com/szagoruyko/pytorchviz) does it. 

What exactly are you trying to accomplish?

**Comment by xcwanAndy** [comment]
Hi @Varal7 , thanks for you suggestions.

I want to achieve this goal: after every batch finishes its forward pass, I want to offload all its tensors (input, intermediate, and output) and the generated computation graph onto CPU. After all batches finish forwarding, I want to put those objects (computation graph and tensors) back to GPU for backward.

`torch.autograd.graph.save_on_cpu()` really helps me a lot. But I think it only helps offload the intermediate and output tensors to CPU, right? So is it possible to offload the input tensors and computation graph to CPU as well?

Maybe offloading computation graph is a bit difficult? So I've tried to additionally offload the input tensors. However, as I tried to put the `input.cpu` back to GPU, I find I cannot put it back to the right place of computation graph for gradient. Is there a way to solve this issue? This may help me a lot as well.

Thanks!

**Comment by albanD** [comment]
Hi,

Just to clarify, the "computation graph" is just a small graph structure that is very small and always in RAM. So you don't need to worry about it here.
The save_on_cpu() will make sure that everything that is needed to run the backward pass (and thus is saved by autograd) is moved to cpu and back.
So I don't think you need to do anything beyond that.

The very first input to the network, you can do whatever you want with it. But if it is needed for backward, then the autograd will have saved it on its own (and offloaded it to cpu properly if you run in save_on_cpu).

**Comment by knagrecha** [comment]
Hi, one issue I have with the save_on_cpu function is that prevent us from restoring intermediates to an alternate GPU - they can only be restored to the original device. For example, if I run a forward pass on GPU 0, transfer the intermediates to CPU, send the graph to GPU 1, restoring intermediates still sends them to GPU 0! Is there a way to create a save_on_cpu hook with more parameters?

**Comment by albanD** [comment]
The save_on_cpu function is 15 lines of python so you can pretty easily change it to do what you want.
https://github.com/pytorch/pytorch/blob/1ac519e6b5f53c02f78d9bae26411affd49328af/torch/autograd/graph.py#L115-L132

In your case, you can add another device argument to the function and have the unpack send to that device instead of the saved one.

Note though that "send the graph to GPU 1" is not something that we officially support and so there might be other parts of the state that are hard-coded to be on GPU 0. So I am not 100% sure this will work.

**Comment by R-N** [comment]
> The save_on_cpu() will make sure that everything that is needed to run the backward pass (and thus is saved by autograd) is moved to cpu and back. 

Hello. My training uses two models that works in two steps. First model produces some tensors in several batches, then second model scores the combination of all batches. The backward pass is also done in two steps using torch.autograd on second model output to compute the gradient wrt tensors produced by first model before actually doing backward pass on those.

So, forward pass is (not actual code)
```
ys = []
for xi in x:
    yi = model1(xi)
    ys.append(yi)
y = torch.cat(ys, dim=0)
z = model2(y)
```

And backward pass is (not actual code) (yes I'm not updating weight of second model as it had been pretrained)
```
dzdy = torch.autograd.grad(inputs=y, outputs=z)[0]
y.backward(gradient=dzdy)
```

So, I have a few questions:
1. Using save_on_cpu, intermediate tensors of the first model forward pass will not be moved to GPU when the torch.autograd.grad is called. Is this correct?
2. Will save_on_cpu load tensors to GPU one by one as needed as it traverses the graph backwards, or will it load everything that the backward pass needs at once?
3. Can the backward pass be done in batch just like the forward pass of first model, to further reduce the GPU memory requirement?
```
n = dzdy.shape[0]
dzdys = [dzdy[i:i+batch] for i in range(0, n, batch)]
for yi, dzdyi in zip(ys, dzdys):
    yi.backward(gradient=dzdyi)
```
